startup/40-xspress3.py

```diff
@@ -0,0 +1,489 @@
+    roi_data = Cpt(PluginBase, 'ROIDATA:')
+    channel1 = Cpt(Xspress3Channel, 'C1_', channel_num=1, read_attrs=['rois'])
+    channel2 = Cpt(Xspress3Channel, 'C2_', channel_num=2, read_attrs=['rois'])
+    channel3 = Cpt(Xspress3Channel, 'C3_', channel_num=3, read_attrs=['rois'])
+    channel4 = Cpt(Xspress3Channel, 'C4_', channel_num=4, read_attrs=['rois'])
+    # create_dir = Cpt(EpicsSignal, 'HDF5:FileCreateDir')
+
+    hdf5 = Cpt(Xspress3FileStoreFlyable, 'HDF5:',
+               read_path_template='/nsls2/xf07bm/data/x3m/%Y/%m/%d/',
+               root='/nsls2/xf07bm/data/',
+               write_path_template='/nsls2/xf07bm/data/x3m/%Y/%m/%d/',
+               )
+
+    def __init__(self, prefix, *, configuration_attrs=None, read_attrs=None,
+                 **kwargs):
+        if configuration_attrs is None:
+            configuration_attrs = ['external_trig', 'total_points',
+                                   'spectra_per_point', 'settings',
+                                   'rewindable']
+        if read_attrs is None:
+            read_attrs = ['channel1', 'channel2', 'channel3', 'channel4', 'hdf5']
+        super().__init__(prefix, configuration_attrs=configuration_attrs,
+                         read_attrs=read_attrs, **kwargs)
+        self.set_channels_for_hdf5()
+        # self.create_dir.put(-3)
+
+        self._asset_docs_cache = deque()
+        self._datum_counter = None
+
+    def stop(self):
+        ret = super().stop()
+        self.hdf5.stop()
+        return ret
+
+    def stage(self):
+        if self.spectra_per_point.get() != 1:
+            raise NotImplementedError(
+                "multi spectra per point not supported yet")
+        ret = super().stage()
+        self._datum_counter = itertools.count()
+        return ret
+
+    def unstage(self):
+        self.settings.trigger_mode.put(0)  # 'Software'
+        super().unstage()
+        self._datum_counter = None
+
+    def complete(self, *args, **kwargs):
+        for resource in self.hdf5._asset_docs_cache:
+            self._asset_docs_cache.append(('resource', resource[1]))
+
+        self._datum_ids = []
+
+        num_frames = xs.hdf5.num_captured.get()
+
+        for frame_num in range(num_frames):
+            datum_id = '{}/{}'.format(self.hdf5._resource_uid, next(self._datum_counter))
+            datum = {'resource': self.hdf5._resource_uid,
+                     'datum_kwargs': {'frame': frame_num},
+                     'datum_id': datum_id}
+            self._asset_docs_cache.append(('datum', datum))
+            self._datum_ids.append(datum_id)
+
+        return NullStatus()
+
+    def collect(self):
+        # TODO: try to separate it from the xspress3 class
+        collected_frames = self.settings.array_counter.get()
+        dpb_sec_values = np.array(dpb_sec.get(), dtype='float128')[:collected_frames * 2: 2]
+        dpb_nsec_values = np.array(dpb_nsec.get(), dtype='float128')[:collected_frames * 2: 2]
```

Comment:
Discussed it with @danielballan and @tacaswell, agreed to use these values as integers, and store them in `data` as separate columns.

The timestamps stores in databroker would be the timestamp of the reading of these waveforms from the IOC. Need to use `.get_with_metadata(...)`:
```suggestion
        dpb_nsec_values = np.array(dpb_nsec.get_with_metadata(), dtype='float128')[:collected_frames * 2: 2]
```

Reply:
This is still on our plate. Right now we are fine with losing the precision when summing seconds and nanoseconds (I think it's the current approach with pizza-boxes, when the timestamp is constricted from the seconds and nanoseconds from the resulted data file.)